src/state.py

```diff
@@ -40,6 +41,62 @@ class Protocol(str, enum.Enum):
     HTTPS = "https"
 
 
+def _validate_hostname_value(value: str) -> str:
+    """Validate the value as a hostname.
+
+    Validation performed:
+    - The hostname must be of length 255 or below.
+    - The hostname must be consist of a certain characters.
+
+    Args:
+        value: The value to validate.
+
+    Raises:
+        ValueError: The validation failed.
+
+    Returns:
+        The value after validation.
+    """
+    if len(value) > 255:
+        raise ValueError("Hostname cannot be longer than 255")
+
+    valid_segment = re.compile(r"(?!-)[A-Z\d-]{1,63}(?<!-)$", re.IGNORECASE)
+    for segment in value.split("."):
+        if valid_segment.fullmatch(segment) is None:
+            raise ValueError(
+                (
+                    "Each Hostname segment must be less than 64 in length, and consist of "
+                    "alphanumeric and hyphen"
+                )
+            )
+
+    return value
+
+
+def _validate_path_value(value: str) -> str:
+    """Validate the value as a path.
+
+    Validation performed:
+    - The path is only consist of allowed characters.
+
+    Args:
+        value: The value to validate.
+
+    Raises:
+        ValueError: The validation failed.
+
+    Returns:
+        The value after validation.
+    """
+    # This are the valid characters for path in addition to `/`:
+    # a-z A-Z 0-9 . - _ ~ ! $ & ' ( ) * + , ; = : @
+    # https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
```

Comment:
Should we put these in the docstring? So that they would come up in the docs easily instead of just saying `allowed characters`.